File: apps/pocket.prompts-app/modules/remote-command/ios/RemoteCommandModule.swift
import ExpoModulesCore
import MediaPlayer
import AVFoundation

/// Native module that captures AirPod / Bluetooth headphone remote commands
/// (play, pause, toggle, next track) via MPRemoteCommandCenter and bridges
/// them to React Native as events.
///
/// To receive commands, the app must be the "now playing" app. We achieve
/// this by playing a nearly-silent audio loop and setting NowPlayingInfo.
public class RemoteCommandModule: Module {
  private var audioPlayer: AVAudioPlayer?
  private var isEnabled = false

  public func definition() -> ModuleDefinition {
    Name("RemoteCommand")

    Events("onRemoteCommand")

    Function("enable") { () -> Void in
      guard !self.isEnabled else { return }
      self.isEnabled = true
      self.setupAudioSession()
      self.setupRemoteCommands()
      self.setupNowPlaying()
    }

    Function("disable") { () -> Void in
      self.teardown()
    }

    Function("isEnabled") { () -> Bool in
      return self.isEnabled
    }

    OnDestroy {
      self.teardown()
    }
  }

  // MARK: - Audio session

  private func setupAudioSession() {
    let session = AVAudioSession.sharedInstance()
    do {
      try session.setCategory(
        .playAndRecord,
        options: [.defaultToSpeaker, .allowBluetooth, .allowBluetoothA2DP, .mixWithOthers]
      )
      try session.setActive(true, options: [])
    } catch {
      print("[RemoteCommand] audio session setup failed: \(error)")
    }

    // re-enable after interruptions (e.g. phone call, Siri)
    NotificationCenter.default.addObserver(
      forName: AVAudioSession.interruptionNotification,
      object: nil,
      queue: .main
    ) { [weak self] notification in
      guard let info = notification.userInfo,
            let typeRaw = info[AVAudioSessionInterruptionTypeKey] as? UInt,
            let type = AVAudioSession.InterruptionType(rawValue: typeRaw) else { return }

      if type == .ended {
        self?.resumeSilentPlayback()
      }
    }
  }

  // MARK: - Remote commands

  private func setupRemoteCommands() {
    let center = MPRemoteCommandCenter.shared()

    // single AirPod tap → togglePlayPause
    center.togglePlayPauseCommand.isEnabled = true
    center.togglePlayPauseCommand.addTarget { [weak self] _ in
      self?.sendEvent("onRemoteCommand", ["command": "togglePlayPause"])
      return .success
    }

    center.playCommand.isEnabled = true
    center.playCommand.addTarget { [weak self] _ in
      self?.sendEvent("onRemoteCommand", ["command": "play"])
      return .success
    }

    center.pauseCommand.isEnabled = true
    center.pauseCommand.addTarget { [weak self] _ in
      self?.sendEvent("onRemoteCommand", ["command": "pause"])
      return .success
    }

    // double tap (AirPods Pro: squeeze-twice) → nextTrack
    center.nextTrackCommand.isEnabled = true
    center.nextTrackCommand.addTarget { [weak self] _ in
      self?.sendEvent("onRemoteCommand", ["command": "nextTrack"])
      return .success
    }

    // triple tap → previousTrack
    center.previousTrackCommand.isEnabled = true
    center.previousTrackCommand.addTarget { [weak self] _ in
      self?.sendEvent("onRemoteCommand", ["command": "previousTrack"])
      return .success
    }
  }

  // MARK: - Now Playing (claims "now playing" status so remote commands arrive)

  private func setupNowPlaying() {
    var info = [String: Any]()
    info[MPMediaItemPropertyTitle] = "pocket.prompts"
    info[MPMediaItemPropertyArtist] = "ready"
    info[MPNowPlayingInfoPropertyIsLiveStream] = true
    MPNowPlayingInfoCenter.default().nowPlayingInfo = info

    startSilentPlayback()
  }

  /// Play a generated silent WAV to claim "now playing" status.
  /// Without this, MPRemoteCommandCenter won't deliver events.
  private func startSilentPlayback() {
    guard let url = silentAudioURL() else { return }
    do {
      audioPlayer = try AVAudioPlayer(contentsOf: url)
      audioPlayer?.numberOfLoops = -1   // loop forever
      audioPlayer?.volume = 0.01        // nearly silent
      audioPlayer?.play()
    } catch {
      print("[RemoteCommand] silent playback failed: \(error)")
    }
  }

  private func resumeSilentPlayback() {
    guard isEnabled else { return }
    try? AVAudioSession.sharedInstance().setActive(true, options: [])
    if audioPlayer?.isPlaying == false {
      audioPlayer?.play()
    }
  }

  /// Generate a minimal 1-second silent WAV file in the temp directory.
  /// 8 kHz mono 8-bit PCM — tiny and sufficient to hold "now playing" status.
  private func silentAudioURL() -> URL? {
    let url = FileManager.default.temporaryDirectory.appendingPathComponent("pp_silence.wav")
    if FileManager.default.fileExists(atPath: url.path) { return url }

    let sampleRate: UInt32 = 8000
    let numSamples = sampleRate          // 1 second
    let bitsPerSample: UInt16 = 8
    let numChannels: UInt16 = 1
    let dataSize = numSamples * UInt32(numChannels) * UInt32(bitsPerSample / 8)

    var d = Data()

    // RIFF header
    d.append(contentsOf: "RIFF".utf8)
    appendLE(&d, UInt32(36 + dataSize))
    d.append(contentsOf: "WAVE".utf8)

    // fmt subchunk
    d.append(contentsOf: "fmt ".utf8)
    appendLE(&d, UInt32(16))         // subchunk size
    appendLE(&d, UInt16(1))          // PCM format
    appendLE(&d, numChannels)
    appendLE(&d, sampleRate)
    appendLE(&d, sampleRate * UInt32(numChannels) * UInt32(bitsPerSample / 8)) // byte rate
    appendLE(&d, UInt16(numChannels * bitsPerSample / 8))                       // block align
    appendLE(&d, bitsPerSample)

    // data subchunk — 8-bit silence is 128 (center of unsigned range)
    d.append(contentsOf: "data".utf8)
    appendLE(&d, dataSize)
    d.append(Data(repeating: 128, count: Int(dataSize)))

    try? d.write(to: url)
    return url
  }

  private func appendLE<T: FixedWidthInteger>(_ data: inout Data, _ value: T) {
    var le = value.littleEndian
    data.append(Data(bytes: &le, count: MemoryLayout<T>.size))
  }

  // MARK: - Teardown

  private func teardown() {
    guard isEnabled else { return }
    isEnabled = false

    let center = MPRemoteCommandCenter.shared()
    center.togglePlayPauseCommand.removeTarget(nil)
    center.playCommand.removeTarget(nil)
    center.pauseCommand.removeTarget(nil)
    center.nextTrackCommand.removeTarget(nil)
    center.previousTrackCommand.removeTarget(nil)

    audioPlayer?.stop()
    audioPlayer = nil

    MPNowPlayingInfoCenter.default().nowPlayingInfo = nil
    NotificationCenter.default.removeObserver(self)
  }
}

File: apps/pocket.prompts-app/modules/remote-command/android/src/main/java/expo/modules/remotecommand/RemoteCommandModule.kt
package expo.modules.remotecommand

import android.content.Context
import android.content.Intent
import android.media.AudioAttributes
import android.media.AudioFocusRequest
import android.media.AudioManager
import android.media.MediaPlayer
import android.media.session.MediaSession
import android.media.session.PlaybackState
import android.os.Build
import android.view.KeyEvent

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

import java.io.File
import java.io.FileOutputStream

/**
 * Android native module that captures Bluetooth headphone media button events
 * (play/pause, next, previous) via MediaSession and bridges them to React
 * Native as events — mirror of the iOS RemoteCommandModule.
 *
 * To receive media button events, the app must be the active "now playing"
 * app. We achieve this by:
 *   1. Creating a MediaSession with an active PlaybackState
 *   2. Requesting audio focus
 *   3. Playing a nearly-silent audio loop
 *
 * Pixel Buds Pro mapping:
 *   single tap    → KEYCODE_MEDIA_PLAY_PAUSE → "togglePlayPause"
 *   (double tap   → KEYCODE_MEDIA_NEXT      → "nextTrack" — if configured)
 *   long press    → Google Assistant (not capturable)
 */
class RemoteCommandModule : Module() {
  private var mediaSession: MediaSession? = null
  private var mediaPlayer: MediaPlayer? = null
  private var audioFocusRequest: AudioFocusRequest? = null
  private var isModuleEnabled = false

  override fun definition() = ModuleDefinition {
    Name("RemoteCommand")

    Events("onRemoteCommand")

    Function("enable") {
      if (isModuleEnabled) return@Function
      isModuleEnabled = true
      setupMediaSession()
      startSilentPlayback()
    }

    Function("disable") {
      teardown()
    }

    Function("isEnabled") {
      isModuleEnabled
    }

    OnDestroy {
      teardown()
    }
  }

  // ── helpers ──────────────────────────────────────────────────────

  private fun emitCommand(command: String) {
    sendEvent("onRemoteCommand", mapOf("command" to command))
  }

  // ── media session ────────────────────────────────────────────────

  private fun setupMediaSession() {
    val context = appContext.reactContext ?: return

    mediaSession = MediaSession(context, "PocketPrompts").apply {
      // intercept raw media button key events for maximum control
      setCallback(object : MediaSession.Callback() {
        override fun onMediaButtonEvent(mediaButtonIntent: Intent): Boolean {
          val event = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            mediaButtonIntent.getParcelableExtra(Intent.EXTRA_KEY_EVENT, KeyEvent::class.java)
          } else {
            @Suppress("DEPRECATION")
            mediaButtonIntent.getParcelableExtra(Intent.EXTRA_KEY_EVENT)
          }

          if (event == null || event.action != KeyEvent.ACTION_DOWN) return false

          when (event.keyCode) {
            // all play/pause variants → togglePlayPause (matches iOS single-tap behavior)
            KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE,
            KeyEvent.KEYCODE_HEADSETHOOK,
            KeyEvent.KEYCODE_MEDIA_PLAY,
            KeyEvent.KEYCODE_MEDIA_PAUSE -> {
              emitCommand("togglePlayPause")
              return true
            }
            KeyEvent.KEYCODE_MEDIA_NEXT -> {
              emitCommand("nextTrack")
              return true
            }
            KeyEvent.KEYCODE_MEDIA_PREVIOUS -> {
              emitCommand("previousTrack")
              return true
            }
          }
          return false
        }
      })

      // set playback state to PLAYING so media buttons route to this session
      setPlaybackState(
        PlaybackState.Builder()
          .setState(PlaybackState.STATE_PLAYING, 0, 1f)
          .setActions(
            PlaybackState.ACTION_PLAY or
            PlaybackState.ACTION_PAUSE or
            PlaybackState.ACTION_PLAY_PAUSE or
            PlaybackState.ACTION_SKIP_TO_NEXT or
            PlaybackState.ACTION_SKIP_TO_PREVIOUS
          )
          .build()
      )

      isActive = true
    }
  }

  // ── silent audio playback ────────────────────────────────────────

  private fun startSilentPlayback() {
    val context = appContext.reactContext ?: return
    val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager

    // request transient audio focus with ducking — this lowers the volume
    // of other audio (music, podcasts) instead of pausing it, so the user's
    // media keeps playing while we hold "now playing" status for headphone
    // button capture. matches the iOS module's .mixWithOthers behavior.
    audioFocusRequest = AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK)
      .setAudioAttributes(
        AudioAttributes.Builder()
          .setUsage(AudioAttributes.USAGE_MEDIA)
          .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
          .build()
      )
      .setOnAudioFocusChangeListener { focusChange ->
        // if we lose focus temporarily, try to resume when we get it back
        if (focusChange == AudioManager.AUDIOFOCUS_GAIN && isModuleEnabled) {
          mediaPlayer?.let { if (!it.isPlaying) it.start() }
        }
      }
      .build()

    audioManager.requestAudioFocus(audioFocusRequest!!)

    // generate + play a 1-second silent WAV loop
    val silentFile = getOrCreateSilentWav(context) ?: return
    try {
      mediaPlayer = MediaPlayer().apply {
        setDataSource(silentFile.absolutePath)
        setAudioAttributes(
          AudioAttributes.Builder()
            .setUsage(AudioAttributes.USAGE_MEDIA)
            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
            .build()
        )
        isLooping = true
        setVolume(0.01f, 0.01f) // nearly silent
        prepare()
        start()
      }
    } catch (e: Exception) {
      android.util.Log.e("RemoteCommand", "silent playback failed: ${e.message}")
    }
  }

  // ── silent WAV generator ─────────────────────────────────────────

  /**
   * Generate a minimal 1-second silent WAV file in the cache directory.
   * 8 kHz mono 8-bit PCM — same format as the iOS module's silentAudioURL().
   * 8-bit unsigned PCM silence = 128 (0x80).
   */
  private fun getOrCreateSilentWav(context: Context): File? {
    val file = File(context.cacheDir, "pp_silence.wav")
    if (file.exists()) return file

    try {
      val sampleRate = 8000
      val numSamples = sampleRate      // 1 second
      val bitsPerSample = 8
      val numChannels = 1
      val dataSize = numSamples * numChannels * (bitsPerSample / 8)

      FileOutputStream(file).use { out ->
        // RIFF header
        out.write("RIFF".toByteArray())
        writeLE32(out, 36 + dataSize)
        out.write("WAVE".toByteArray())

        // fmt subchunk
        out.write("fmt ".toByteArray())
        writeLE32(out, 16)             // subchunk size
        writeLE16(out, 1)              // PCM format
        writeLE16(out, numChannels)
        writeLE32(out, sampleRate)
        writeLE32(out, sampleRate * numChannels * (bitsPerSample / 8)) // byte rate
        writeLE16(out, numChannels * (bitsPerSample / 8))              // block align
        writeLE16(out, bitsPerSample)

        // data subchunk — 8-bit unsigned silence = 128 (0x80)
        out.write("data".toByteArray())
        writeLE32(out, dataSize)
        out.write(ByteArray(dataSize) { 0x80.toByte() })
      }

      return file
    } catch (e: Exception) {
      android.util.Log.e("RemoteCommand", "failed to create silent wav: ${e.message}")
      return null
    }
  }

  private fun writeLE16(out: FileOutputStream, value: Int) {
    out.write(value and 0xFF)
    out.write((value shr 8) and 0xFF)
  }

  private fun writeLE32(out: FileOutputStream, value: Int) {
    out.write(value and 0xFF)
    out.write((value shr 8) and 0xFF)
    out.write((value shr 16) and 0xFF)
    out.write((value shr 24) and 0xFF)
  }

  // ── teardown ─────────────────────────────────────────────────────

  private fun teardown() {
    if (!isModuleEnabled) return
    isModuleEnabled = false

    mediaSession?.isActive = false
    mediaSession?.release()
    mediaSession = null

    mediaPlayer?.stop()
    mediaPlayer?.release()
    mediaPlayer = null

    val context = appContext.reactContext
    if (context != null && audioFocusRequest != null) {
      val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
      audioManager.abandonAudioFocusRequest(audioFocusRequest!!)
    }
    audioFocusRequest = null
  }
}
